zebra-network/src/isolated.rs

```diff
@@ -1,62 +1,69 @@
-    Service,
+    ServiceExt,
 };
 
-use zebra_chain::chain_tip::NoChainTip;
+use zebra_chain::{chain_tip::NoChainTip, parameters::Network};
 
 use crate::{
     peer::{self, ConnectedAddr, HandshakeRequest},
     peer_set::ActiveConnectionCounter,
     BoxError, Config, Request, Response,
 };
 
-/// Use the provided TCP connection to create a Zcash connection completely
-/// isolated from all other node state.
+#[cfg(feature = "tor")]
+pub(crate) mod tor;
+
+#[cfg(test)]
+mod tests;
+
+/// Creates a Zcash peer connection using the provided data stream.
+/// This connection is completely isolated from all other node state.
 ///
-/// The connection pool returned by `init` should be used for all requests that
+/// The connection pool returned by [`init`](zebra_network::init)
+/// should be used for all requests that
 /// don't require isolated state or use of an existing TCP connection. However,
 /// this low-level API is useful for custom network crawlers or Tor connections.
 ///
 /// In addition to being completely isolated from all other node state, this
 /// method also aims to be minimally distinguishable from other clients.
 ///
+/// SECURITY TODO: check if the timestamp field can be zeroed, to remove another distinguisher (#3300)
+///
 /// Note that this method does not implement any timeout behavior, so callers may
 /// want to layer it with a timeout as appropriate for their application.
 ///
 /// # Inputs
 ///
-/// - `conn`: an existing TCP connection to use. Passing an existing TCP
-/// connection allows this method to be used with clearnet or Tor transports.
-///
-/// - `user_agent`: a valid BIP14 user-agent, e.g., the empty string.
+/// - `network`: the Zcash [`Network`] used for this connection.
 ///
-/// # Bug
+/// - `data_stream`: an existing data stream. This can be a non-anonymised TCP connection,
+///                  or a Tor client [`DataStream`].
 ///
-/// `connect_isolated` only works on `Mainnet`, see #1687.
-pub fn connect_isolated(
-    conn: TcpStream,
+/// - `user_agent`: a valid BIP14 user-agent, e.g., the empty string.
+pub fn connect_isolated<AsyncReadWrite>(
+    network: Network,
+    data_stream: AsyncReadWrite,
     user_agent: String,
-) -> impl Future<
-    Output = Result<
-        BoxService<Request, Response, Box<dyn std::error::Error + Send + Sync + 'static>>,
-        Box<dyn std::error::Error + Send + Sync + 'static>,
-    >,
-> {
+) -> impl Future<Output = Result<BoxService<Request, Response, BoxError>, BoxError>>
+where
+    AsyncReadWrite: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+{
+    let config = Config {
+        network,
+        ..Config::default()
+    };
+
     let handshake = peer::Handshake::builder()
-        .with_config(Config::default())
+        .with_config(config)
         .with_inbound_service(tower::service_fn(|_req| async move {
-            Ok::<Response, Box<dyn std::error::Error + Send + Sync + 'static>>(Response::Nil)
+            Ok::<Response, BoxError>(Response::Nil)
```

Comment:
Optional suggestion: Not sure if it improves things or not:

```suggestion
            Ok::<_, BoxError>(Response::Nil)
```